lib/rubocop/cop/rails/index_true.rb

```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Rails
+      # Check that index: true is not being used in add_column or change_column statement.
+      # This option functions as expected in a create_table statement, but silently
+      # fails in other migration statements.
+      #
+      # @example
+      #   # bad
+      #   add_column :booking_templates, :booking_id, :integer, index: true
+      #
+      #   # good
+      #   add_column :booking_templates, :booking_id, :integer
+      #
+      #   # Remember to add the index separately
+      #   add_index :booking_templates, :booking_id
+      #
+      class IndexTrue < Cop
+        MSG = '`index: true` does not work in an `add_column` or `change_column` ' \
+        'method. Please use `add_index :table, :column`.'.freeze
+
+        def_node_matcher :no_index_true, <<-PATTERN
+          (send nil {:add_column :change_column} _table _column _type (hash $...))
+        PATTERN
+
+        def on_send(node)
+          no_index_true(node) do |args|
+            add_offense(node, :expression, format(MSG)) if index_option_passed?(args)
+          end
+        end
+
+        def index_option_passed?(args)
+          args.first.node_parts.first.children.first == :index
```

Comment:
@bbatsov Thanks for all the helpful feedback! Appreciate how active you are here. Made two additional commits to address the issues:
https://github.com/bbatsov/rubocop/pull/4396/commits/cbbaaad19ab2000c5662f608f98f8cf233cd40ba
https://github.com/bbatsov/rubocop/pull/4396/commits/7ee38ee9df02c95fd9534f4c4fbe50ed46c95b73